### PR TITLE
[SYSTEMML-1743] Create lightweight SystemML artifact

### PR DESCRIPTION
**Pull request by deroneriksson**
Create a lightweight SystemML jar file based on the execution of
a particular set of code, such as JMLC. In addition to generating
a minimal jar file that corresponds to this code, automatically
generate maven assembly dependency sets that can be used to build
a 'lite' jar that can be activated via a 'lite' maven profile.
This essentially creates an minimum-size uberjar that can perform
tasks such as scoring with JMLC, where distributed operations on
Spark and Hadoop are not desired. Note that this potentially addresses
the underlying requirements of situations such as that needed by
[SYSTEMML-453].

### CODE REVIEW

**Comment by deroneriksson** [comment]
Lightweight jar can be built using:
```
mvn clean package -P distribution,lite
```

@mboehm7 Please let me know what you think of this approach and if other examples (such as those in BuildLiteExecution.java) should be supported.

**Comment by deroneriksson** [comment]
Also, note that the artifact created by the maven assembly includes all log4j classes, commons-math3 classes, and SystemML classes.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1771/

File: src/main/java/org/apache/sysml/utils/lite/BuildLite.java [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+		System.out.println("\nCreating " + liteJarLocation + " file");
+		ClassLoader cl = BuildLite.class.getClassLoader();
+
+		Manifest mf = new Manifest();
+		Attributes attr = mf.getMainAttributes();
+		attr.putValue("" + Attributes.Name.MANIFEST_VERSION, "1.0");
+
+		File file = new File(liteJarLocation);
+		FileOutputStream fos = new FileOutputStream(file);
+		JarOutputStream jos = new JarOutputStream(fos, mf);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/utils/lite/BuildLite.java]: potential resource leak - please use try/finally or try-with-resources

File: src/main/java/org/apache/sysml/utils/lite/BuildLite.java [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+	 *             if an IOException occurs
+	 * @throws ClassNotFoundException
+	 *             if a ClassNotFoundException occurs
+	 */
+	private static List<String> getAllClassesInJar(Class<?> classInJarFile) throws IOException, ClassNotFoundException {
+		List<String> classPathNames = new ArrayList<String>();
+		String jarLocation = classInJarFile.getProtectionDomain().getCodeSource().getLocation().getPath();
+		File f = new File(jarLocation);
+		FileInputStream fis = new FileInputStream(f);
+		JarArchiveInputStream jais = new JarArchiveInputStream(fis);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/utils/lite/BuildLite.java]: see above

File: src/main/java/org/apache/sysml/utils/lite/BuildLite.java [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+					totalMatches++;
+					found = true;
+					String match = m.group(1);
+					String matchClass = additionalPackage + match;
+					uniqueMatches.add(matchClass);
+					// System.out.println("FOUND " + matchClass);
+				}
+				if (!found) {
+					// System.out.println("NOT FOUND " + s);
+				}
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/utils/lite/BuildLite.java]: please remove those unnecessary code blocks and commented code.

File: src/main/java/org/apache/sysml/utils/lite/BuildLiteExecution.java [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+					continue;
+				}
+				matrix[i][j] = (random.nextDouble() * (max - min) + min);
+			}
+		}
+		return matrix;
+	}
+
+	public static void displayMatrix(double[][] matrix) {
+		System.out.println("Matrix size:" + matrix.length + "x" + matrix[0].length);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/utils/lite/BuildLiteExecution.java]: you might want to replace this with DataConverter.toString(DataConverter.convertToMatrixBlock(matrix)).

File: src/main/java/org/apache/sysml/utils/lite/LiteCheck.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		try {
+			is = LiteCheck.class.getResourceAsStream("/" + BuildLite.LITE_JAR_IDENTIFIER_FILE);
+			return (is != null);
+		} finally {
+			if (is != null) {
+				try {
+					is.close();
+				} catch (IOException e) {
+				}
+			}
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/utils/lite/LiteCheck.java]: Please use try-with-resources or IOUtilFunctions.closeSilently()

**Comment by mboehm7** [comment]
Regarding log4j, it would be great if we could also package a default log4j.properties into this jar; otherwise users won't see warnings from the compiler, etc or would need to include a custom log4j.properties into their classpath. Furthermore, I ran into this error (which, however, did not raise an exception):

 `java.lang.ClassNotFoundException: org.apache.hadoop.log.metrics.EventCounter
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at org.apache.log4j.helpers.Loader.loadClass(Loader.java:179)
        at org.apache.log4j.helpers.OptionConverter.instantiateByClassName(OptionConverter.java:320)
        at org.apache.log4j.helpers.OptionConverter.instantiateByKey(OptionConverter.java:121)
        at org.apache.log4j.PropertyConfigurator.parseAppender(PropertyConfigurator.java:664)
        at org.apache.log4j.PropertyConfigurator.parseCategory(PropertyConfigurator.java:647)
        at org.apache.log4j.PropertyConfigurator.configureRootCategory(PropertyConfigurator.java:544)
        at org.apache.log4j.PropertyConfigurator.doConfigure(PropertyConfigurator.java:440)
        at org.apache.log4j.PropertyConfigurator.doConfigure(PropertyConfigurator.java:476)
        at org.apache.log4j.helpers.OptionConverter.selectAndConfigure(OptionConverter.java:471)
        at org.apache.log4j.LogManager.<clinit>(LogManager.java:125)
        at org.slf4j.impl.Log4jLoggerFactory.getLogger(Log4jLoggerFactory.java:66)
        at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:270)
        at org.apache.commons.logging.impl.SLF4JLogFactory.getInstance(SLF4JLogFactory.java:155)
        at org.apache.commons.logging.LogFactory.getLog(LogFactory.java:290)
        at org.apache.sysml.hops.OptimizerUtils.<clinit>(OptimizerUtils.java:60)
        at org.apache.sysml.api.DMLScript$DMLOptions.<init>(DMLScript.java:129)
        at org.apache.sysml.api.DMLScript$DMLOptions.<clinit>(DMLScript.java:138)
        at org.apache.sysml.api.DMLScript.<clinit>(DMLScript.java:161)
        at <some testclass>
log4j:ERROR Could not instantiate appender named "EventCounter"
`

**Comment by deroneriksson** [comment]
List of updates:
```
Shrink lite.xml size by over 50%.
Add MRJobConfiguration.USE_BINARYBLOCK_SERIALIZATION check.
Switch isLite/remotePar checks in InfrastructureAnalyzer.
Update BuildLite IO to use try-with-resources.
Remove commented code in BuildLite.
Switch displayMatrix to use DataConverter.
Update LiteCheck isLite to use try-with-resources.
Cache isLite results for performance.
Add default log4j.properties to root of artifact.
Add EventCounter class to lite artifact.
Exclude lite jar from python artifact if built at same time.
Generate entire dependencySets section from BuildLite.
Automatically update lite.xml dependencySets if lite.xml is found.
Allow commons-math3 option for all or only required classes.
Switch BuildLiteExecution output to utilize log4j.
```

Note that I tried switching to `MRJobConfiguration.addBinaryBlockSerializationFramework(_rJob)` but with this minimal lite jar I saw `Could not find a serializer for the Key class: 'org.apache.sysml.runtime.matrix.data.MatrixIndexes'`, so I kept the existing `MRConfigurationNames.IO_SERIALIZATIONS` call within the `USE_BINARYBLOCK_SERIALIZATION` check.

By handling the lite.xml differently for commons-logging and commons-math3, I was able to shrink the size of lite.xml by over 50%. WRT switching to calling BuildLiteExecution from the lite profile, I would say artifact maintentance would become very difficult (WRT issues such as licensing, included files, etc) since it would mean going from essentially our project's standard maven assembly technique for building our verified artifacts to a fairly convoluted programmatic technique. I think the lite.xml file used with the maven-assembly-plugin helps standardize how the artifact is built, which makes our release validation process easier. A side benefit of decoupling the actual BuildLiteExecution execution from the lite profile means that the artifact takes a few seconds to build rather than a few minutes.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1788/

**Comment by deroneriksson** [comment]
@mboehm7 When you have time, please let me know what you think of the latest updates.

**Comment by deroneriksson** [comment]
I'll go ahead and merge this tomorrow if there are no objections.

**Comment by bertholdreinwald** [comment]
can you please hold off a few more days. Thanks.


Regards,
Berthold Reinwald
IBM Almaden Research Center
office: (408) 927 2208; T/L: 457 2208
e-mail: reinwald@us.ibm.com

**Comment by deroneriksson** [comment]
ping @bertholdreinwald

**Comment by deroneriksson** [comment]
ping @bertholdreinwald

**Comment by bertholdreinwald** [comment]
sorry for the delay. 

I tried it on a bare bone machine. Got below class not found. Checked the lite jar. Several org.apache.hadoop.util classes are included but GenericOptionsParser is missing. Should it be included?

Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/hadoop/util/GenericOptionsParser
	at org.apache.sysml.api.DMLScript.main(DMLScript.java:230)
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.util.GenericOptionsParser
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 1 more

**Comment by deroneriksson** [comment]
@bertholdreinwald Thanks for having a look at this. It looks like you're running SystemML using DMLScript (`at org.apache.sysml.api.DMLScript.main(DMLScript.java:230)`) rather than using the JMLC API, which is the focus of the lite jar. Please try with the JMLC API.

**Comment by deroneriksson** [comment]
Rebased with latest code. Also updated l2svm predict input param due to a change introduced in the last month.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2030/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2099/

**Pull request closed without merging**